orchagent/neighorch.cpp

```diff
@@ -1135,20 +1145,131 @@ void NeighOrch::doVoqSystemNeighTask(Consumer &consumer)
 
 bool NeighOrch::addInbandNeighbor(string alias, IpAddress ip_address)
 {
-    //For "port" type inband, the inband reachability info syncing can be done through static
-    //configureation or CHASSIS_APP_DB sync (this function)
+    //Add neighbor record in SAI without adding host route for local inband to avoid route
+    //looping for packets destined to the Inband interface if the Inband is port type
+
+    if(gIntfsOrch->isRemoteSystemPortIntf(alias))
+    {
+        //Remote Inband interface. Skip
+        return true;
+    }
+
+    sai_status_t status;
+    MacAddress inband_mac = gMacAddress;
+
+    sai_object_id_t rif_id = gIntfsOrch->getRouterIntfsId(alias);
+    if (rif_id == SAI_NULL_OBJECT_ID)
+    {
+        SWSS_LOG_INFO("Failed to get rif_id for %s", alias.c_str());
+        return false;
+    }
+
+    //Make the object key
+    sai_neighbor_entry_t neighbor_entry;
+    neighbor_entry.rif_id = rif_id;
+    neighbor_entry.switch_id = gSwitchId;
+    copy(neighbor_entry.ip_address, ip_address);
+
+    vector<sai_attribute_t> neighbor_attrs;
+    sai_attribute_t attr;
+    attr.id = SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS;
+    memcpy(attr.value.mac, inband_mac.getMac(), 6);
+    neighbor_attrs.push_back(attr);
+
+    if(!addVoqEncapIndex(alias, ip_address, neighbor_attrs))
+    {
+        return false;
+    }
+
+    //No host route for neighbor of the Inband IP address
+    attr.id = SAI_NEIGHBOR_ENTRY_ATTR_NO_HOST_ROUTE;
+    attr.value.booldata = true;
+    neighbor_attrs.push_back(attr);
 
-    //For "vlan" type inband, the inband reachability info syncinng can be ARP learning of other
-    //asics inband or static configuration or through CHASSIS_APP_DB sync (this function)
+    status = sai_neighbor_api->create_neighbor_entry(&neighbor_entry, static_cast<uint32_t>(neighbor_attrs.size()), neighbor_attrs.data());
```

Comment:
my question what is the use of creating this information from SAI perspective. How this will be used later ?

Reply:
Just a brief info about inband interface and recycle port: Inband interface is for cpu-to-cpu communication. In this PR we use recycle port facility for inband interface. Think of recycle port as any other network port but the egress side is looped back to the ingress. Since host is reached via inband interface for all the protocol communications we need to add the self neighbor on the inband interface. This is like any other neighbor on the front panel port. Since the neighbor resides within  the host of the asic itself, we use the recycle facility to loop back the packet and add neighbor to consume the packet. 
After adding the neighbor we inform the neighbor to other asics in the same we do for any neighbor on any other front panel port. Other asics program this neighbor on the system port of inband interface of this asic, When remote asics sends packet to this asic, they inject packet on their cpu port and everything else is similar to regular routing handled in the hardware.
For more information on this please refer to the section 2.5.1 of the voq HLD https://github.com/Azure/SONiC/blob/master/doc/voq/voq_hld.md.

Reply:
@vganesan-nokia thanks i had gone through this document before posting the question. I am not able to follow why there is need to create self neighbor for recycle port.  Let me know what i am missing in below :
LC1 CPU want to send packet destined to LC2 Inband IP and need to be consumed by LC2 CPU. 
For this my understanding is 

- Packet from LC1 CPU will read LC1 Recycle Port via NetDev
- Packet will do Host Lookup in LC1 Ingress ASIC and will send packet to LC2 Recylce Port
- Packet will reach LC2 Recycle Port and looped back into Ingress ASIC of LC2
- LC2 will have /32 route for its Inband IP and will go to LC2 CPU via Netdev of LC2 Recycle port.

So I am not sure where the self neighbor entry is coming into picture and how SAI will use it.

Reply:
> @vganesan-nokia thanks i had gone through this document before posting the question. I am not able to follow why there is need to create self neighbor for recycle port. Let me know what i am missing in below :
> LC1 CPU want to send packet destined to LC2 Inband IP and need to be consumed by LC2 CPU.
> For this my understanding is
> 
> * Packet from LC1 CPU will reach LC1 Recycle Port via NetDev
```[vganesan-nokia] >>> Here, the packet injection is on asic's cpu port with packet's destination information pointing to egress of recycle port.```
> * Packet will get looped back do Host Lookup in LC1 Ingress ASIC and will send packet to LC2 Recylce Port
```[vganesan-nokia] >>> For here, when the packet is looped back, i.e., when packet is received on ingress of the same LC1 asic it has to be directed to router block```. **For this consumption to happen we need the neighbor info of self on the recycle port**. ```Once packet is in routing operation block, the routing operation determines the destination information i.e., the system port of inband interface of LC2```
> * Packet will reach LC2 Recycle Port and looped back into Ingress ASIC of LC2
```[vganesan-nokia] >>> Here from fabric, the packet reaches the egress of the recycle port in LC2. This is looped back to ingress where the host IP trap rule sends packet to CPU.``` 
> * LC2 will have /32 route for its Inband IP and will go to LC2 CPU via Netdev of LC2 Recycle port.
> 
> So I am not sure where the self neighbor entry is coming into picture and how SAI will use it.

Reply:
The inband port IP address of the "my-asic" is the next-hop in the routing tables of the "other-asics" - for routes that point to "my-asic" host IP stack. So "my-asic" needs to create this neighbor and let the other asics know about this neighbor and its encap index. This will allow the other asics to host traffic destined to "my-asic"me via the inband port of "my-asic".

This is the only reason to create this neighbor. Please note that for this neighbor we do not create a host route in SAI. The neighbor attribute SAI_NEIGHBOR_ENTRY_ATTR_NO_HOST_ROUTE is used for this purpose. We do that becuase -
1. The IP2ME trap rules can already send the packet to the CPU once it comes around to the ingress of the recyle port.
2. If we have the host-route for this neighbor - it loop around the recyle port instead of being trapped to the CPU.

Reply:
> 
> 
> sorry i forgot. Can you please tells us the reason why we are creating Neighbor record on SAI for the local Inband/Recycle Port interface ? How this will get used ? Which section of design document cover the need for doing this ?

[skeesara-nokia] This is needed so that we get an encap index that the other asics can use it for the case where the next-hop is the inband port IP of my-asic

Reply:
As discussed please update HLD with the information of L2Header (Src/Dest Mac) being used for the above packat flow.

Reply:
Done. A PR has been raised with the requested update.

https://github.com/Azure/SONiC/pull/776